File: 303. Range Sum Query - Immutable/solution.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
  let sum = 0;
  this.dp = nums.map(num => (sum += num));
};

/**
 * @param {number} i
 * @param {number} j
 * @return {number}
 */
NumArray.prototype.sumRange = function(i, j) {
  if (i === 0) return this.dp[j];
  return this.dp[j] - this.dp[i - 1];
};

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = Object.create(NumArray).createNew(nums)
 * var param_1 = obj.sumRange(i,j)
 */

File: 303. Range Sum Query - Immutable/solution.py
class NumArray:

    def __init__(self, nums: List[int]):
        total = 0
        self._sum = [0]
        for n in nums:
            total += n
            self._sum.append(total)

    def sumRange(self, i: int, j: int) -> int:
        return self._sum[j + 1] - self._sum[i]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)
